fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/AndOrValue.java

```diff
@@ -0,0 +1,219 @@
+import com.google.common.collect.Iterables;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A {@link Value} that applies conjunction/disjunction on its boolean children, and if possible, simplifies its boolean children.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class AndOrValue implements BooleanValue {
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("And-Or-Value");
+    @Nonnull
+    protected final String functionName;
+    @Nonnull
+    protected final Value leftChild;
+    @Nonnull
+    protected final Value rightChild;
+    @Nonnull
+    protected final Operator operator;
+
+    private enum Operator {
+        AND,
+        OR
+    }
+
+    /**
+     * Constructs a new instance of {@link AndOrValue}.
+     *
+     * @param functionName The function name.
+     * @param leftChild The left child.
+     * @param rightChild The right child.
+     * @param operator The actual comparison operator.
+     */
+    protected AndOrValue(@Nonnull String functionName,
+                         @Nonnull Operator operator,
+                         @Nonnull Value leftChild,
+                         @Nonnull Value rightChild) {
+        this.functionName = functionName;
+        this.leftChild = leftChild;
+        this.rightChild = rightChild;
+        this.operator = operator;
+    }
+
+    @Nonnull
+    @Override
+    public Iterable<? extends Value> getChildren() {
+        return ImmutableList.of(leftChild, rightChild);
+    }
+
+    @Override
+    public int semanticHashCode() {
+        return PlanHashable.objectsPlanHash(PlanHashKind.FOR_CONTINUATION, BASE_HASH, functionName, leftChild, rightChild);
+    }
+
+    @Override
+    public int planHash(@Nonnull final PlanHashKind hashKind) {
+        return PlanHashable.objectsPlanHash(hashKind, BASE_HASH, functionName, leftChild, rightChild);
+    }
+
+    @Override
+    public String toString() {
+        return functionName + "(" + leftChild + ", " + rightChild + ")";
+    }
+
+    @Override
+    public int hashCode() {
+        return semanticHashCode();
+    }
+
+    @SuppressWarnings("EqualsWhichDoesntCheckParameterClass")
+    @SpotBugsSuppressWarnings("EQ_UNUSUAL")
+    @Override
+    public boolean equals(final Object other) {
+        return semanticEquals(other, AliasMap.identitiesFor(getCorrelatedTo()));
+    }
+
+    @Nullable
+    @Override
+    public <M extends Message> Object eval(@Nonnull final FDBRecordStoreBase<M> store,
+                                           @Nonnull final EvaluationContext context,
+                                           @Nullable final FDBRecord<M> fdbRecord,
+                                           @Nullable final M message) {
+        final Object leftResult = leftChild.eval(store, context, fdbRecord, message);
+
+        if (leftResult == null) {
+            return null;
```

Comment:
`UNKNOWN AND FALSE` is `FALSE`. `UNKNOWN OR TRUE` is `TRUE`.

Reply:
done, thanks for pointing it out.